<a href="https://colab.research.google.com/github/safakhurshid836-bot/Artificial_Intelligence_Lab/blob/main/Bank_Data_Science_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# https://archive.ics.uci.edu/dataset/222/bank+marketing

import pandas as pd
df = pd.read_csv('/content/bank-full.csv', sep=';')
display(df.head(100))

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,36,management,married,tertiary,no,101,yes,yes,unknown,5,may,426,1,-1,0,unknown,no
96,55,blue-collar,married,secondary,no,383,no,no,unknown,5,may,287,1,-1,0,unknown,no
97,60,retired,married,tertiary,no,81,yes,no,unknown,5,may,101,1,-1,0,unknown,no
98,39,technician,married,secondary,no,0,yes,no,unknown,5,may,203,1,-1,0,unknown,no


In [2]:
from google.colab import ai
response = ai.generate_text("What is the capital of France?")

In [4]:
print('using .head()')
display(df.head())
print('using .info()')
df.info(20)
print('using .describe()')
df.describe()

using .head()


,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no


using .info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   45211 non-null  object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB
using .describe()


,age,balance,day,duration,campaign,pdays,previous
count,45211.000000,45211.000000,45211.000000,45211.000000,45211.000000,45211.000000,45211.000000
mean,40.936210,1362.272058,15.806419,258.163080,2.763841,40.197828,0.580323
std,10.618762,3044.765829,8.322476,257.527812,3.098021,100.128746,2.303441
min,18.000000,-8019.000000,1.000000,0.000000,1.000000,-1.000000,0.000000
25%,33.000000,72.000000,8.000000,103.000000,1.000000,-1.000000,0.000000
50%,39.000000,448.000000,16.000000,180.000000,2.000000,-1.000000,0.000000
75%,48.000000,1428.000000,21.000000,319.000000,3.000000,-1.000000,0.000000
max,95.000000,102127.000000,31.000000,4918.000000,63.000000,871.000000,275.000000


In [5]:
missing_values = df.isnull().sum()  #missing_value aik variable hai
display(missing_values[missing_values > 0])
print(missing_values)    #kitni values missing hain
df.rename(columns={"y":"subscriptions"}, inplace=True)  #to rename a certain coloumn and heading
display(df.head(100)) #to display 100 rows of data

,0


age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64


,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,subscriptions
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,36,management,married,tertiary,no,101,yes,yes,unknown,5,may,426,1,-1,0,unknown,no
96,55,blue-collar,married,secondary,no,383,no,no,unknown,5,may,287,1,-1,0,unknown,no
97,60,retired,married,tertiary,no,81,yes,no,unknown,5,may,101,1,-1,0,unknown,no
98,39,technician,married,secondary,no,0,yes,no,unknown,5,may,203,1,-1,0,unknown,no


Let's create a sample DataFrame with missing values to illustrate `ffill` and `bfill`.

In [6]:
import pandas as pd
import numpy as np

sample_data = {'A': [1, 2, np.nan, 4, 5],
               'B': [np.nan, 7, 8, np.nan, 10],
               'C': [11, 12, 13, 14, np.nan]}
sample_df = pd.DataFrame(sample_data)

print("Original DataFrame with missing values:")
display(sample_df)

Original DataFrame with missing values:


,A,B,C
0,1.0,NaN,11.0
1,2.0,7.0,12.0
2,NaN,8.0,13.0
3,4.0,NaN,14.0
4,5.0,10.0,NaN


### Using Forward Fill (`ffill`)

In [9]:
df_ffill = sample_df.ffill()
print("DataFrame after forward fill (ffill):")
display(df_ffill)

DataFrame after forward fill (ffill):


,A,B,C
0,1.0,NaN,11.0
1,2.0,7.0,12.0
2,2.0,8.0,13.0
3,4.0,8.0,14.0
4,5.0,10.0,14.0


### Using Backward Fill (`bfill`)

In [15]:
df_bfill = sample_df.bfill()
print("DataFrame after backward fill (bfill):")
display(df_bfill)

DataFrame after backward fill (bfill):


,A,B,C
0,1.0,7.0,11.0
1,2.0,7.0,12.0
2,4.0,8.0,13.0
3,4.0,10.0,14.0
4,5.0,10.0,NaN


Other common methods for handling missing values include:

*   **Filling with a specific value**: `df.fillna(value)` (e.g., `df.fillna(0)` to replace NaNs with 0).
*   **Filling with the mean, median, or mode**: `df['column'].fillna(df['column'].mean())`.
*   **Dropping rows/columns with missing values**: `df.dropna()`.

In [ ]:
#if there is any value missing in the data set we can use the forward fill and backward fill methods
# Commonly used in data analysis tools like Python's Pandas library (using df.fillna(method='ffill') or bfill).

# Task
Preprocess the `df` DataFrame by performing the following steps: check for and remove duplicate rows, identify categorical and numerical features, apply one-hot encoding to categorical features, standardize numerical features using `StandardScaler`, and then display the first few rows and information about the resulting preprocessed DataFrame. Finally, summarize all the preprocessing operations that were performed.

## Check for Duplicates

### Subtask:
Identify and remove any duplicate rows in the DataFrame to ensure data uniqueness.


**Reasoning**:
First, I'll identify and display the number of duplicate rows in the DataFrame to understand the extent of the problem. Then, I'll remove these duplicates from the DataFrame to ensure data uniqueness, and finally, I'll display the new shape of the DataFrame to confirm the duplicates have been successfully removed.



In [21]:
print(f"Number of duplicate rows before removal: {df.duplicated().sum()}")
df.drop_duplicates(inplace=True)
print(f"DataFrame shape after removing duplicates: {df.shape}")

Number of duplicate rows before removal: 0
DataFrame shape after removing duplicates: (45211, 17)


## Identify Categorical and Numerical Features

**Reasoning**:
The next step in preprocessing is to identify which columns are categorical and which are numerical. This is a crucial step for applying appropriate preprocessing techniques later, such as one-hot encoding for categorical features and standardization for numerical features. I will use a code block to automatically classify columns based on their data types.



In [22]:
categorical_features = df.select_dtypes(include='object').columns
numerical_features = df.select_dtypes(include=['int64', 'float64']).columns

print(f"Categorical Features ({len(categorical_features)}):\n{list(categorical_features)}")
print(f"Numerical Features ({len(numerical_features)}):\n{list(numerical_features)}")

Categorical Features (10):
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome', 'subscriptions']
Numerical Features (7):
['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']


## Apply One-Hot Encoding

**Reasoning**:
To prepare the categorical features for machine learning models, I will apply one-hot encoding using `pd.get_dummies`. This converts each categorical variable into a new set of binary variables, which is suitable for many algorithms.



In [18]:
df_encoded = pd.get_dummies(df, columns=categorical_features, drop_first=True)
print("DataFrame after one-hot encoding:")
display(df_encoded.head())

DataFrame after one-hot encoding:


,age,balance,day,duration,campaign,pdays,previous,job_blue-collar,job_entrepreneur,job_housemaid,...,month_jun,month_mar,month_may,month_nov,month_oct,month_sep,poutcome_other,poutcome_success,poutcome_unknown,subscriptions_yes
0,58,2143,5,261,1,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,False
1,44,29,5,151,1,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,False
2,33,2,5,76,1,-1,0,False,True,False,...,False,False,True,False,False,False,False,False,True,False
3,47,1506,5,92,1,-1,0,True,False,False,...,False,False,True,False,False,False,False,False,True,False
4,33,1,5,198,1,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,False


## Standardize Numerical Features

**Reasoning**:
Now that one-hot encoding has been applied, I will standardize the numerical features to ensure that they contribute equally to the model training, preventing features with larger values from dominating. I will use `StandardScaler` from `sklearn.preprocessing` for this purpose and then display the first few rows of the modified DataFrame.



In [19]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
df_encoded[numerical_features] = scaler.fit_transform(df_encoded[numerical_features])

print("DataFrame after standardizing numerical features:")
display(df_encoded.head())

DataFrame after standardizing numerical features:


,age,balance,day,duration,campaign,pdays,previous,job_blue-collar,job_entrepreneur,job_housemaid,...,month_jun,month_mar,month_may,month_nov,month_oct,month_sep,poutcome_other,poutcome_success,poutcome_unknown,subscriptions_yes
0,1.606965,0.256419,-1.298476,0.011016,-0.569351,-0.411453,-0.25194,False,False,False,...,False,False,True,False,False,False,False,False,True,False
1,0.288529,-0.437895,-1.298476,-0.416127,-0.569351,-0.411453,-0.25194,False,False,False,...,False,False,True,False,False,False,False,False,True,False
2,-0.747384,-0.446762,-1.298476,-0.707361,-0.569351,-0.411453,-0.25194,False,True,False,...,False,False,True,False,False,False,False,False,True,False
3,0.571051,0.047205,-1.298476,-0.645231,-0.569351,-0.411453,-0.25194,True,False,False,...,False,False,True,False,False,False,False,False,True,False
4,-0.747384,-0.447091,-1.298476,-0.233620,-0.569351,-0.411453,-0.25194,False,False,False,...,False,False,True,False,False,False,False,False,True,False


**Reasoning**:
After standardizing numerical features, I will now display the information about the final preprocessed DataFrame to show its structure and data types after all the transformations.



In [20]:
print("Information about the preprocessed DataFrame:")
df_encoded.info()

Information about the preprocessed DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 43 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   age                  45211 non-null  float64
 1   balance              45211 non-null  float64
 2   day                  45211 non-null  float64
 3   duration             45211 non-null  float64
 4   campaign             45211 non-null  float64
 5   pdays                45211 non-null  float64
 6   previous             45211 non-null  float64
 7   job_blue-collar      45211 non-null  bool   
 8   job_entrepreneur     45211 non-null  bool   
 9   job_housemaid        45211 non-null  bool   
 10  job_management       45211 non-null  bool   
 11  job_retired          45211 non-null  bool   
 12  job_self-employed    45211 non-null  bool   
 13  job_services         45211 non-null  bool   
 14  job_student          45211 non-null  boo

## Summary of Preprocessing Operations

The following preprocessing operations were performed on the `df` DataFrame:

1.  **Duplicate Removal**: Checked for and removed any duplicate rows. In this case, no duplicate rows were found, so the DataFrame's shape remained unchanged.
2.  **Feature Identification**: Categorical features were identified (type `object`) and numerical features (types `int64`, `float64`) were separated.
3.  **One-Hot Encoding**: Categorical features were converted into numerical format using one-hot encoding with `pd.get_dummies`, and the first column for each categorical feature was dropped to avoid multicollinearity.
4.  **Numerical Feature Standardization**: Numerical features were standardized using `StandardScaler` from `sklearn.preprocessing`. This scaled the numerical values to have a mean of 0 and a standard deviation of 1.

## Summary of Preprocessing Operations

The following preprocessing operations were performed on the `df` DataFrame:

1.  **Duplicate Removal**: Checked for and removed any duplicate rows. In this case, no duplicate rows were found, so the DataFrame's shape remained unchanged.
2.  **Feature Identification**: Categorical features were identified (type `object`) and numerical features (types `int64`, `float64`) were separated.
3.  **One-Hot Encoding**: Categorical features were converted into numerical format using one-hot encoding with `pd.get_dummies`, and the first column for each categorical feature was dropped to avoid multicollinearity.
4.  **Numerical Feature Standardization**: Numerical features were standardized using `StandardScaler` from `sklearn.preprocessing`. This scaled the numerical values to have a mean of 0 and a standard deviation of 1.

## Summary:

### Data Analysis Key Findings

*   **Duplicate Rows**: No duplicate rows were found in the DataFrame. The shape of the DataFrame remained `(45211, 17)` after the duplicate removal step.
*   **Feature Identification**:
    *   10 categorical features were identified: `job`, `marital`, `education`, `default`, `housing`, `loan`, `contact`, `month`, `poutcome`, `subscriptions`.
    *   7 numerical features were identified: `age`, `balance`, `day`, `duration`, `campaign`, `pdays`, `previous`.
*   **One-Hot Encoding**: Categorical features were successfully one-hot encoded using `pd.get_dummies` with `drop_first=True`. This process increased the number of columns from 17 to 43.
*   **Numerical Feature Standardization**: Numerical features were standardized using `StandardScaler`, ensuring they have a mean of approximately 0 and a standard deviation of 1.
*   **Final DataFrame Structure**: The preprocessed DataFrame (`df_encoded`) contains 45211 entries and 43 columns, consisting of 7 `float64` columns (standardized numerical features) and 36 `bool` columns (one-hot encoded categorical features). All columns were confirmed to be non-null.

### Insights or Next Steps

*   The `df_encoded` DataFrame is now prepared for machine learning model training, as categorical features have been converted to a numerical format and numerical features have been scaled to a consistent range.
*   The next step could involve splitting the `df_encoded` DataFrame into training and testing sets, followed by selecting and training a predictive model.
